### PR TITLE
test(manifestation): replace `run_future()` with `TestContext`

### PR DESCRIPTION
**Pull request by rami3l**
As a continuation of https://github.com/rust-lang/rustup/pull/3367 and #3856, this PR makes all tests in `dist::manifestation::tests` that rely on `run_future()` async, thus the latter can be fully removed.

### CODE REVIEW

File: src/dist/manifestation/tests.rs [hunk 1]
```diff
@@ -2,13 +2,15 @@
...
 #![allow(clippy::type_complexity)]
 
-use std::collections::HashMap;
-use std::env;
-use std::fs;
-use std::path::Path;
-use std::str::FromStr;
-use std::sync::Arc;
-use std::{cell::Cell, future::Future};
+use std::{
```
Comment (djc) [inline src/dist/manifestation/tests.rs:5]: Can you make a separate commit here that changes the formatting?

File: src/dist/manifestation/tests.rs [hunk 2]
```diff
@@ -31,6 +33,7 @@ use crate::{
 const SHA256_HASH_LEN: usize = 64;
 
 /// Temporary thunk to support asyncifying from underneath.
+#[deprecated]
```
Comment (djc) [inline src/dist/manifestation/tests.rs]: I think we should squash this commit and the next one; they're not logically independent (this just adds unused code which would trigger warnings when running `cargo check` at this commit) and it's easier to review the new code when it's in the same commit as the old code being removed.

**Review by djc — COMMENTED** [review]
Awesome! Would like some tweaks before I review this in a bit more detail...

**Comment by rami3l** [comment]
@djc I don't know how you'd like to proceed with your reviews, but just FYI I find this PR a perfect fit for language-aware diffs like SemanticDiff: https://app.semanticdiff.com/gh/rust-lang/rustup/pull/3915/commits

File: src/dist/manifestation/tests.rs [hunk 3]
```diff
@@ -388,108 +349,28 @@ fn rename_component_new() {
...
+    let cx = TestContext::from_dist_server(mock_dist_server, url, GZOnly);
+
+    let adds = [Component::new(
+        "bobo".to_string(),
+        Some(TargetTriple::new("x86_64-apple-darwin")),
+        true,
+    )];
+    // Install the basics from day 1
+    change_channel_date(&cx.url, "nightly", "2016-02-01");
+    cx.update_from_dist(&[], &[], false).await.unwrap();
```
Comment (djc) [inline src/dist/manifestation/tests.rs:361]: Probably not worth doing this now, but in the future, consider splitting the commits by first isolating the `TestContext` type to contain all the arguments given to these closures (generally by the same name) and passing the `TestContext` type to the testing closures as before (and perhaps also attaching the `update_from_dist()` function to that `TestContext` type). That would have made this a lot easier to review.

File: src/dist/manifestation/tests.rs [hunk 4]
```diff
@@ -535,195 +400,205 @@ impl Compressions {
...
-    comps: Compressions,
-    f: &dyn Fn(&Url, &ToolchainDesc, &InstallPrefix, &DownloadCfg<'_>, &temp::Context),
-) {
-    server.write(
-        &[MockManifestVersion::V2],
-        comps.enable_xz(),
-        comps.enable_zst(),
-    );
+impl TestContext {
+    fn new(edit: Option<&dyn Fn(&str, &mut MockChannel)>, comps: Compressions) -> Self {
```
Comment (djc) [inline src/dist/manifestation/tests.rs:414]: I think this could benefit from a more descriptive name. I feel like now the relation between `new()` and `from_dist_server()` is the opposite from what I expect (`new()` is generally lower level than other constructors, but in this case it is arguably higher level.)

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
> @djc I don't know how you'd like to proceed with your reviews, but just FYI I find this PR a perfect fit for language-aware diffs like SemanticDiff: https://app.semanticdiff.com/gh/rust-lang/rustup/pull/3915/commits

Ooohhh, that is pretty awesome. Should try to use that more.

File: src/dist/manifestation/tests.rs (hunk 4 above)
Comment (rami3l) [reply src/dist/manifestation/tests.rs:414]: @djc Hmmm, `setup` used to be a short name and I changed it to `new` simply because of the "more uses = terser name" convention.

That said, what about `Self::from_edit()`? I'm really bad at naming 🤣

Comment (djc) [reply src/dist/manifestation/tests.rs:414]: Hmm, on review I feel like `new()` is probably reasonable here based on how it is used, even if it is a little counterintuitive to me. Let's just merge it?
